lib/ret_web/controllers/api/v1/avatar_controller.ex

```diff
@@ -0,0 +1,192 @@
+    conn |> send_resp(401, "")
+  end
+
+  defp create_or_update(conn, params, avatar) do
+    account = conn |> Guardian.Plug.current_resource()
+
+    files_to_promotoe =
+      case params["files"] do
+        nil ->
+          %{}
+
+        files ->
+          files
+          |> Enum.map(fn {k, v} -> {String.to_atom(k), List.to_tuple(v)} end)
+          |> Enum.into(%{})
+      end
+
+    owned_file_results = Storage.promote(files_to_promotoe, account)
+
+    promotion_error =
+      owned_file_results |> Map.values() |> Enum.filter(&(elem(&1, 0) == :error)) |> Enum.at(0)
+
+    case promotion_error do
+      nil ->
+        owned_files =
+          owned_file_results |> Enum.map(fn {k, {:ok, file}} -> {k, file} end) |> Enum.into(%{})
+
+        parent_avatar =
+          if params["parent_avatar_id"] do
+            Repo.get_by(Avatar, avatar_sid: params["parent_avatar_id"])
+          end
+
+        {result, avatar} =
+          avatar
+          |> Avatar.changeset(account, owned_files, parent_avatar, params)
+          |> Repo.insert_or_update()
+
+        avatar = avatar |> Repo.preload(@file_columns)
+
+        case result do
+          :ok ->
+            conn |> render("create.json", avatar: avatar)
+
+          :error ->
+            conn |> send_resp(422, "invalid avatar")
+        end
+
+      {:error, :not_found} ->
+        conn |> send_resp(404, "no such file(s)")
+
+      {:error, :not_allowed} ->
+        conn |> send_resp(401, "")
+    end
+  end
+
+  defp collapse_avatar_files(%{parent_avatar: nil} = avatar),
+    do: avatar |> Map.take(@file_columns)
+
+  defp collapse_avatar_files(%{parent_avatar: parent} = avatar) do
+    Map.merge(collapse_avatar_files(parent), avatar |> Map.take(@file_columns), fn _k, v1, v2 ->
+      if is_nil(v2), do: v1, else: v2
+    end)
+  end
+
+  defp get_avatar(avatar_sid) do
+    avatar =
+      Avatar
+      |> Repo.get_by(avatar_sid: avatar_sid)
+      |> Repo.preload([@file_columns ++ [:parent_avatar, :account]])
+  end
+
+  def show(conn, %{"id" => avatar_sid}) do
+    conn |> show(avatar_sid |> get_avatar())
+  end
+
+  def show(conn, nil = _avatar) do
+    conn |> send_resp(404, "")
+  end
+
+  def show(conn, %Avatar{} = avatar) do
+    conn |> render("show.json", avatar: avatar)
+  end
+
+  def show_gltf(conn, %{"id" => avatar_sid}) do
+    conn |> show_gltf(Avatar |> Repo.get_by(avatar_sid: avatar_sid))
+  end
+
+  def show_gltf(conn, nil = _avatar) do
+    conn |> send_resp(404, "")
+  end
+
+  def show_gltf(conn, %Avatar{} = avatar) do
+    # TODO we ideally don't need to be featching the OwnedFiles until after we collapse them
+    avatar_files =
+      avatar
+      |> Avatar.load_parents(@file_columns)
+      |> Map.from_struct()
+      |> collapse_avatar_files()
+
+    case Storage.fetch(avatar_files.gltf_owned_file) do
+      {:ok, %{"content_type" => content_type, "content_length" => content_length}, stream} ->
+        gltf =
+          stream
+          |> Enum.join("")
```

Comment:
You can try 
```
stream
|> Stream.concat
|> Stream.run
|> Poison.decode!
```

if that works it will result in less blocking and i believe also less copying -- will just end up concatenating everything streamed through into a materialized enum i think

Reply:
(because of the `run` it will block the call to decode until the stream is finished, which is less than ideal, so you could try it without the `run` too, but i think `Poison` expects an `iolist`)

Reply:
Hmm messed with this for a bit but couldn't get another solution to work